### ISSUE TITLE
Purpose of the challenge-response mechanism?

### ISSUE THREAD

**Issue by ibeckermayer** [ben-cohen/transient_token#4]
Hi @ben-cohen, this is a very creative idea for authenticating with xrdp. I'm trying to understand the details of this implementation, but am getting caught up on one aspect: I don't quite understand what the challenge response mechanism is buying you here.

As i understand it, the user executes `get_transient_token` over ssh and that generates them a token of the form `TTK<uid>:<pid>:<challenge><response>`. They then use that token as the password of an xrdp session.

That password is passed into your custom PAM module which parses it and then sends the `<challenge>` to the socket kicked off by `get_transient_token`, and if that `<challenge>` matches the one that was generated then the socket passes back `<response>`. The PAM module then checks that `<response>` matches the response it expects from the password-token and if it does, it logs the user in. 

That all makes sense to me, however I'm not quite groking what that challenge-response is doing. Would it not be functionally equivalent to just send the whole token to the socket, and have it check the entire token string matches, and then just send back a `"PASS"` or `"FAIL"` string rather than the `<response>`?

Based on your README it seems as though this may be how things used to work:

https://github.com/ben-cohen/transient_token/blob/6f591f82c23c0bfaa1b144798ec7180b8c0506ca/README#L33-L34

was there some particular reason you added the challenge-response? Is it setting things up for an extra layer of security that's TODO? Or am I just missing some way in which this is actually better than the "functionally equivalent" system I proposed above?

**Comment by ben-cohen** [comment]
Hi @ibeckermayer,  yes, you are right that the response is not necessary.  Thank you for pointing that out!

When I shortened the token in commit [8623ad69e](https://github.com/ben-cohen/transient_token/commit/8623ad69e69c559303daab626ecb00da7c2f512d), I changed the README from "expects to receive the correct response" to the wording you quoted.  I think I intended to remove the response, perhaps in a second commit, but forgot.

(I wrote this so I could run a Linux GUI desktop in AWS and connect using RDP, and I had fun writing some tools I would have used, also including a plugin to forward the SSH agent over RDP. :)  But in the end I chose a different workflow and stopped working on it.)

I removed the response and tidied things up in the 4 commits that I pushed just now.
